### ISSUE TITLE
build: Add ARMv7 to the supported builds

### ISSUE THREAD

**Issue by overload** [rclone/rclone#748]
version: rclone-v1.33-freebsd-arm
hardware: Hardkernel ODROID-C1
OS: FreeBSD 11.0-CURRENT

ERROR:
"runtime: this system has multiple CPUs and must use
atomic synchronization instructions. Recompile using GOARM=7."

I think must be recompiled for multi-CPU

**Comment by ncw** [comment]
I should probably make an ARMv7 release as well as the ARMv6 which is the default... 

In the meantime you should be able to compile from source

**Comment by ncw** [comment]
I compiled a version from source with

    GOOS=freebsd GOARCH=arm GOARM=7 go build

Can you see if it works?

http://pub.rclone.org/rclone-v1.33-28-gf2eeb43-freebsd-armv7.zip

**Comment by overload** [comment]
Il 28/09/2016 17:23, Nick Craig-Wood ha scritto:
> Can you see if it works?
Yes, it seems to work
-- 

Regards,
Andrea Cerquozzi

**Comment by overload** [comment]
Il 28/09/2016 14:09, Nick Craig-Wood ha
      scritto:
    
    I should probably make an ARMv7 release as well as the
      ARMv6 which is the default... 
    ok: ARMv7 work fine for ODROID-C1+
    
    Please note that also Raspberry Pi 2 model B use ARMv7 architecture,
    and now ODROID-C2 & R.Pi3 use ARMv8 CPU
    http://www.hardkernel.com/main/products/prdt_info.php
    
    
    Thanks.
    -- 
      Regards,
        Andrea Cerquozzi
      OVERLOAD Home Entertainment
        www.overload.it
      P.S. siamo da poco anche su facebook, non
          perdere i nostri aggiornamenti e curiosità,
          diventa
            fan ...conviene!

**Comment by ncw** [comment]
Note that ARMv6 code runs fine on ARMv7 processors under linux.  It is a limitation freebsd (or the go runtime running under freebsd - I'm not sure which) that ARMv6 code doesn't run on ARMv7 processors.

https://golang.org/src/runtime/os_freebsd_arm.go

That said an ARMv7 version would almost certainly run faster on the new raspberry Pi.

ARMv8 is known as arm64 and I've added that to the release process already - you can find a beta here, though I think the Raspberry Pi runs in ARMv7 mode by default.

http://pub.rclone.org/v1.33-24-gbfe6f29%CE%B2/

I'll try to figure out a sensible way of adding ARMv7 to the release process.

**Comment by animosity22** [comment]
@overload  - is this taken care of now or still something outstanding?

**Comment by overload** [comment]
I'm using an old version copiled with GOARM=7, it works.
When I try latest version, I receive the same message:
```
runtime: this system has multiple CPUs and must use
atomic synchronization instructions. Recompile using GOARM=7.
```

**Comment by ncw** [comment]
I think we need to make a decision here

We could either

1. Upgrade the ARM build to v7 from v6
    - this will cause old processors to stop working

Or

2. Build an ARMv7 and an ARMv6 release

My preference would be for 1. as it is less releases to maintain (it is also a 1 line patch!) however I don't know how many ARMv6 processors are out there. @overload do you have an opinion on that? Do we need to build an official ARMv6 build?

**Comment by overload** [comment]
I think ARMv6 is a very old processor:
https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures

the first versions of raspberry pi use ARMv6, they could still be widespread
from wikipedia: "The Raspberry Pi Zero and Zero W use the same Broadcom BCM2835 SoC as the first generation Raspberry Pi" so ARMv6

Raspberry pi 2 use ARMv7
Raspberry pi 3/4 use ARMv8

**Comment by calisro** [comment]
As much as I wanted to say drop v6, There are a lot of raspberry pi v6 use cases out there.  They are capable.  I personally use one for things because I had some laying around.  I can compile my own but if I use them, i'm sure others do.  It'll likely break people.

**Comment by ncw** [comment]
OK let's support both v6 and v7

Currently the downloads look like this.

Suggestions for what to call the ARM32 bit rows? I know users will have problems working out which they need.

Maybe `ARMv7 - 32 Bit` and `ARMv6 - 32 Bit` would that work?

![image](https://user-images.githubusercontent.com/536803/83351015-6d91f700-a338-11ea-8b44-5d176051a682.png)

**Comment by overload** [comment]
ARM 64 bit is ARMv8, while v7 is an ARM 32bit with multicore processor
can anyone test if the compiled version for ARMv7 also works on ARMv8? It might work.

**Comment by overload** [comment]
ARMv6 - 32 Bit
ARMv7 - 32 Bit (multi)
ARMv8 - 64 Bit (multi)

**Comment by ncw** [comment]
> can anyone test if the compiled version for ARMv7 also works on ARMv8? It might work

**Comment by ncw** [comment]
I had a go at this here

https://beta.rclone.org/branch/v1.52.0-019-gb625ded5-fix-748-arm-v7-beta/ (uploaded in 15-30 mins)

Can you give the relevant binary a go and see if it works for you - thanks!

**Comment by overload** [comment]
```
runtime: this CPU has no floating point hardware, so it cannot run
this GOARM=7 binary. Recompile using GOARM=5.
```
on ARMv7... is it possibile? the old v1.33 compiled with GOARM=7 works

**Comment by ncw** [comment]
What CPU are you running on? I didn't think there were very many ARMv7 without floating point? Also shouldn't your OS be implementing softfloat for you? Which OS are you using?

> on ARMv7... is it possibile? the old v1.33 compiled with GOARM=7 works

Compiled which with go version?

**Comment by overload** [comment]
processor Amlogic S905, OS FreeBSD 11

compiled by ncw on 28 Sep 2016 (see post)
http://pub.rclone.org/rclone-v1.33-28-gf2eeb43-freebsd-armv7.zip (unavailable)

**Comment by ncw** [comment]
Does FreeBSD not have floating point emulator support compiled in? I would expect the OS to be emulating the floating point instructions if the chip doesn't support them directly.

**Comment by ncw** [comment]
I've merged this to master now which means there will be an ARMv7 build in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.53
